RandLAPACK/drivers/rl_revd2.hh

```diff
@@ -0,0 +1,240 @@
+
+        // B = Y(R')^-1 - need to transpose R
+        blas::trsm(Layout::ColMajor, Side::Right, Uplo::Upper, Op::NoTrans, Diag::NonUnit, m, k, 1.0, R_dat, k, Y_dat, m);
+
+        //[V, S, ~] = SVD(B)
+        // Although we don't need the right singular vectors, we need to give space for those.
+        // Use R as a buffer for that.
+        lapack::gesdd(Job::SomeVec, m, k, Y_dat, m, S_dat, V_dat, m, R_dat, k);
+
+        // E = diag(S^2)
+        T buf;
+        int64_t r = 0;
+        int i;
+        for(i = 0; i < k; ++i) {
+            buf = std::pow(S[i], 2);
+            E[i] = buf;
+            // r = number of entries in E that are greater than v
+            if(buf > nu)
+                ++r;
+        }
+
+        // Undo regularlization
+        for(i = 0; i < r; ++i)
+            E[i] -=nu;
+
+        std::fill(V.begin() + m * r, V.end(), 0.0);
+
+        // Error estimation
+        // Using the first column of Omega as a buffer for a random vector
+        // To perform the following safely, need to make sure Omega has at least 4 columns
+        Omega_dat = util::upsize(m * 4, this->Omega);
+        RandBLAS::dense::DenseDist  g{.n_rows = m, .n_cols = 1};
+        RandBLAS::dense::fill_buff(Omega_dat, g, state);
+
+        for(int i = 0; i < this->p; ++i) {
+            T g_norm = blas::nrm2(m, Omega_dat, 1);
+            // Compute g = g / ||g|| - we need this because dot product does not take in an alpha
+            std::for_each(Omega_dat, Omega_dat + m, [g_norm](T &entry) { entry /= g_norm;});
+
+            // Compute V'*g / ||g||
+            // Using the second column of Omega as a buffer for matrix-vector product
+            gemv(Layout::ColMajor, Op::Trans, m, k, 1.0, V_dat, m, Omega_dat, 1, 0.0, Omega_dat + m, 1);
+
+
+            // Compute V*E, E diag
+            // Using Y as a buffer for V*E
+            for (int i = 0, j = 0; i < m * k; ++i) {
+                Y[i] = V[i] * E[j];
+                if((i + 1) % m == 0 && i != 0)
+                    ++j;
+            }
+
+            // Compute V*E*V'*g / ||g||
+            // Using the third column of Omega as a buffer for matrix-vector product
+            gemv(Layout::ColMajor, Op::NoTrans, m, k, 1.0, Y_dat, m,Omega_dat + m, 1, 0.0, Omega_dat + 2 * m, 1);
+            // Compute A*g / ||g||
+            // Using the forth column of Omega as a buffer for matrix-vector product
+            gemv(Layout::ColMajor, Op::NoTrans, m, m, 1.0, A_dat, m, Omega_dat, 1, 0.0, Omega_dat + 3 * m, 1);
+            // Compute A*g / ||g|| - V*E*V'*g / ||g||
+            // Result is stored in the 4th column of Omega
+            blas::axpy(m, -1.0, Omega_dat + 2 * m, 1, Omega_dat + 3 * m, 1);
+            // Compute (g / ||g||)' * (A*g / ||g|| - V*E*V'*g / ||g||) - this is our measure for the error
+            err = blas::dot(m, Omega_dat, 1, Omega_dat + 3 * m, 1);	
+            // v_0 <- v
+            std::copy(Omega_dat + 3 * m, Omega_dat + 4 * m, Omega_dat);
+        }
```

Comment:
Have you considered writing a helper function that performs this operation? If you did, then it would be easier to test this code.

Reply:
Yes, i just don't know where to put it. 
It would be too specialized for utils.
I guess i could make it a part of REVD2 class?

Reply:
Hmm. At first, I thought that I wasn't so special. It's just a method for estimating the largest eigenvalue of a PSD matrix. But then I realized that our PSD matrix, ``A - V E V'``, is represented implicitly. That does complicate things.

My suggestion would be to make it a helper function in the ``rl_evd2.hh`` file. No need to attach it to the ``REVD2`` class at this time.

Reply:
Sure, on it.